File: frontend/src/shared/domain/value-objects.ts
// Shared domain value objects
export abstract class ValueObject {
  abstract equals(other: ValueObject): boolean;
  abstract hashCode(): number;
}

export class Money extends ValueObject {
  constructor(
    public readonly amount: number,
    public readonly currency: string = 'EUR'
  ) {
    super();
    if (amount < 0) {
      throw new Error('Money amount cannot be negative');
    }
    if (!currency) {
      throw new Error('Currency cannot be empty');
    }
  }

  equals(other: ValueObject): boolean {
    if (!(other instanceof Money)) return false;
    return this.amount === other.amount && this.currency === other.currency;
  }

  hashCode(): number {
    return this.amount.toString().concat(this.currency).length;
  }

  add(other: Money): Money {
    if (this.currency !== other.currency) {
      throw new Error('Cannot add money with different currencies');
    }
    return new Money(this.amount + other.amount, this.currency);
  }

  multiply(factor: number): Money {
    return new Money(this.amount * factor, this.currency);
  }

  format(): string {
    return new Intl.NumberFormat('es-ES', {
      style: 'currency',
      currency: this.currency,
    }).format(this.amount);
  }
}

export class Email extends ValueObject {
  constructor(public readonly value: string) {
    super();
    if (!this.isValidEmail(value)) {
      throw new Error(`Invalid email format: ${value}`);
    }
  }

  equals(other: ValueObject): boolean {
    if (!(other instanceof Email)) return false;
    return this.value.toLowerCase() === other.value.toLowerCase();
  }

  hashCode(): number {
    return this.value.toLowerCase().length;
  }

  private isValidEmail(email: string): boolean {
    const pattern = /^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$/;
    return pattern.test(email);
  }
}

export class UserId extends ValueObject {
  constructor(public readonly value: string) {
    super();
    if (!value || !value.trim()) {
      throw new Error('UserId cannot be empty');
    }
  }

  equals(other: ValueObject): boolean {
    if (!(other instanceof UserId)) return false;
    return this.value === other.value;
  }

  hashCode(): number {
    return this.value.length;
  }
}

export class PokemonId extends ValueObject {
  constructor(public readonly value: number) {
    super();
    if (value <= 0) {
      throw new Error('PokemonId must be positive');
    }
  }

  equals(other: ValueObject): boolean {
    if (!(other instanceof PokemonId)) return false;
    return this.value === other.value;
  }

  hashCode(): number {
    return this.value;
  }
}

export class CartId extends ValueObject {
  constructor(public readonly value: string) {
    super();
    if (!value || !value.trim()) {
      throw new Error('CartId cannot be empty');
    }
  }

  equals(other: ValueObject): boolean {
    if (!(other instanceof CartId)) return false;
    return this.value === other.value;
  }

  hashCode(): number {
    return this.value.length;
  }
}

File: backend/src/shared/domain/value_objects.py
"""Shared domain value objects."""

from abc import ABC, abstractmethod
from dataclasses import dataclass
from decimal import Decimal
from typing import Any, Self
import re


class ValueObject(ABC):
    """Base class for all value objects."""
    
    @abstractmethod
    def __eq__(self, other: Any) -> bool:
        pass
    
    @abstractmethod
    def __hash__(self) -> int:
        pass


@dataclass(frozen=True)
class Money(ValueObject):
    """Money value object with currency support."""
    
    amount: Decimal
    currency: str = "USD"
    
    def __post_init__(self) -> None:
        if self.amount < 0:
            raise ValueError("Money amount cannot be negative")
        if not self.currency:
            raise ValueError("Currency cannot be empty")
    
    def __eq__(self, other: Any) -> bool:
        if not isinstance(other, Money):
            return False
        return self.amount == other.amount and self.currency == other.currency
    
    def __hash__(self) -> int:
        return hash((self.amount, self.currency))
    
    def add(self, other: "Money") -> "Money":
        """Add two money values with same currency."""
        if self.currency != other.currency:
            raise ValueError("Cannot add money with different currencies")
        return Money(self.amount + other.amount, self.currency)
    
    def multiply(self, factor: int | float) -> "Money":
        """Multiply money by a factor."""
        return Money(self.amount * Decimal(str(factor)), self.currency)


@dataclass(frozen=True)
class Email(ValueObject):
    """Email value object with validation."""
    
    value: str
    
    def __post_init__(self) -> None:
        if not self._is_valid_email(self.value):
            raise ValueError(f"Invalid email format: {self.value}")
    
    def __eq__(self, other: Any) -> bool:
        if not isinstance(other, Email):
            return False
        return self.value.lower() == other.value.lower()
    
    def __hash__(self) -> int:
        return hash(self.value.lower())
    
    @staticmethod
    def _is_valid_email(email: str) -> bool:
        """Validate email format using regex."""
        pattern = r'^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$'
        return bool(re.match(pattern, email))


@dataclass(frozen=True)
class UserId(ValueObject):
    """User identifier value object."""
    
    value: str
    
    def __post_init__(self) -> None:
        if not self.value or not self.value.strip():
            raise ValueError("UserId cannot be empty")
    
    def __eq__(self, other: Any) -> bool:
        if not isinstance(other, UserId):
            return False
        return self.value == other.value
    
    def __hash__(self) -> int:
        return hash(self.value)


@dataclass(frozen=True)
class PokemonId(ValueObject):
    """Pokemon identifier value object."""
    
    value: int
    
    def __post_init__(self) -> None:
        if self.value <= 0:
            raise ValueError("PokemonId must be positive")
    
    def __eq__(self, other: Any) -> bool:
        if not isinstance(other, PokemonId):
            return False
        return self.value == other.value
    
    def __hash__(self) -> int:
        return hash(self.value)


@dataclass(frozen=True)
class CartId(ValueObject):
    """Cart identifier value object."""
    
    value: str
    
    def __post_init__(self) -> None:
        if not self.value or not self.value.strip():
            raise ValueError("CartId cannot be empty")
    
    def __eq__(self, other: Any) -> bool:
        if not isinstance(other, CartId):
            return False
        return self.value == other.value
    
    def __hash__(self) -> int:
        return hash(self.value)
